### DISCUSSION TITLE
Como vocês fazem CI/CD de apps em React?

### DISCUSSION THREAD

**Discussion by wilsonneto-swe** [frontendbr/forum#1577] [Perguntas e Respostas]
Fala galera, 
tudo bem?

Sou programador a mais tempo com .Net e faço pipelines no Azure DevOps para aplicações .net tranquilo... 

Porém gostaria de saber como vocês fazem, quais ferramentas usam, para pipelines de apps em React?
Principalmente referente a ambientes on premise...

Em busca de insights e opiniões.

Abraços.

**Comment by wmartins** [comment]
Tu tem alguma dúvida específica sobre? Achei a pergunta bem ampla, mas vou tentar colocar algumas coisas que já usei para a discussão.

Trabalhei em um projeto onde usávamos Jenkins para fazer o build da aplicação. No nosso caso, fazíamos o build usando um container Docker com todas as dependências (node, npm e algumas outras libs do próprio SO). Pelo que eu me lembre o que possibilitava isso era o [Docker Plugin](https://plugins.jenkins.io/docker-plugin).

Usávamos a [versão open source do Nexus](https://www.sonatype.com/download-oss-sonatype) como gerenciador de artefatos (para pacotes internos de npm e para imagens Docker).

Em um outro projeto trabalhei com Circle CI (não on premise, mas acho que vale para a discussão), mas também usando os builds com Docker. A diferença é que tínhamos que ter alguns comandos de `npm install` e `apt-get` para podermos "acertar" o container de build. Coisa que não era necessária nesse projeto com Jenkins, já que essa imagem já estava _pré buildada_.

---

Em um resumão, normalmente os pipelines em si são bem simples, tendo `npm` e `node` (que tu pode instalar com `nvm` por conveniência), fica bem fácil de automatizar tudo via `npm scripts` ou uma meia dúzia de shell scripts.

Claro que dependendo do teu projeto (se tem server side rendering, por exemplo), talvez vai ter alguma coisa ou outra que vai ter que ser mockada/configurada, mas no geral, falando somente do build _estático_ do React, é bem tranquilo.

**Comment by tcelestino** [comment]
Em um projeto atual, utilizamos também o Jenkins para o pipeline, que a partir de vários docker-compose, que usamos para rodar os testes, para gerar os build do next.js e fazer o upload dos arquivos estáticos para a CDN e após todo processo, subimos a imagem para o nosso serviço de "orquestração" de containers (no caso, o swarm).

Em tese, se não tiver fazendo um SSR (nosso caso), você só precisa ter no seu pipeline o próprio `npm run build` (se tiver utilizando o create-react-app, já tá configurado) e rodar o `npm start` dentro do seu container.

**Comment by wilsonneto-swe** [comment]
Show, @wmartins não tinha uma dúvida específica não, mais pela discussão e opiniões de como fazem.
Aqui por hora fizemos em Azure DevOps e um agent/serviço local em nosso servidor on premise que é windows, assim nem precisamos mandar os arquivos, o próprio agente já baixa e faz os jobs de relesase.

Estamos estruturando um deploy no Jenkins também para vermos ambos funcionando e ver qual vai ser melhor para nós.

@tcelestino legal, aqui vamos testar o Jenkins também.

Abraços

**Comment by ninetails** [comment]
O bom de usar Circle CI/Travis CI é que são ferramentas online. Infelizmente só as usei para projetos open source (repo público) que dá pra usar gratuitamente e o bom é que integram com o GitHub, dá pra deixar _triggers_ quando por exemplo abrem _Pull Request_.

Num dos trampos já cheguei a usar [Buddy](https://buddy.works/), creio que funcione tanto pra CI quanto pra CD. Em anteriores, também usei _Pipelines_ do _GitLab_.

Pra brincar (estudar) cheguei a criar algumas [GitHub Actions](https://github.com/features/actions), o bom é que como ainda tá em _beta_, tá _open_ pra usar. Cheguei a usar pra rodar _npm scripts_ previamente configurados com `lint`, `test` e teste _e2e_. [E configurar foi bem fácil](https://github.com/ninetails/challenge-magalu/blob/develop/.github/workflows/main.yml) mas precisa ter liberado o _beta testing_ do GitHub. Daí ele mostra numa _tab_ do repositório e também configurando o _workflow_ para algum _hook_ como PRs. Por trás sobe um _Docker_, tranquilo. [Deve ser sussegado também criar alguma _action_ para CD e rodar quando criar um _release_.](https://help.github.com/en/github/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onpushpull_requestbranchestags) Mas ressalto: tá em _beta_.
